packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2125,46 +2139,39 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     _lastBottomViewInset = WidgetsBinding.instance!.window.viewInsets.bottom;
   }
 
-  _WhitespaceDirectionalityFormatter? _whitespaceFormatter;
+  late final _WhitespaceDirectionalityFormatter _whitespaceFormatter = _WhitespaceDirectionalityFormatter(textDirection: _textDirection);
 
   void _formatAndSetValue(TextEditingValue value) {
-    _whitespaceFormatter ??= _WhitespaceDirectionalityFormatter(textDirection: _textDirection);
-
     // Check if the new value is the same as the current local value, or is the same
     // as the pre-formatting value of the previous pass (repeat call).
-    final bool textChanged = _value.text != value.text;
-    final bool isRepeat = value == _lastFormattedUnmodifiedTextEditingValue;
-
-    // There's no need to format when starting to compose or when continuing
-    // an existing composition.
-    final bool isComposing = value.composing.isValid;
-    final bool isPreviouslyComposing = _lastFormattedUnmodifiedTextEditingValue?.composing.isValid ?? false;
+    final bool textChanged = _value.text != value.text || _value.composing != value.composing;
 
-    if ((textChanged || (!isComposing && isPreviouslyComposing)) &&
-        widget.inputFormatters != null &&
-        widget.inputFormatters!.isNotEmpty) {
+    if (textChanged) {
       // Only format when the text has changed and there are available formatters.
       // Pass through the formatter regardless of repeat status if the input value is
       // different than the stored value.
-      for (final TextInputFormatter formatter in widget.inputFormatters!) {
-        value = formatter.formatEditUpdate(_value, value);
-      }
+      value = widget.inputFormatters?.fold<TextEditingValue>(
+        value,
+        (TextEditingValue newValue, TextInputFormatter formatter) => formatter.formatEditUpdate(_value, newValue),
+      ) ?? value;
+
       // Always pass the text through the whitespace directionality formatter to
       // maintain expected behavior with carets on trailing whitespace.
-      value = _whitespaceFormatter!.formatEditUpdate(_value, value);
-      _lastFormattedValue = value;
+      value = _whitespaceFormatter.formatEditUpdate(_value, value);
     }
 
-    // Setting _value here ensures the selection and composing region info is passed.
+    // Before triggering the [_value] change notifier, we want to make sure
+    // [_updateRemoteEditingValueIfNeeded] is muted, so we can consolidate the
+    // changes and send them to the engine in one batch.
+    _muteRemoteUpdate = true;
```

Comment:
I think the mute idea might have the potential to create confusing behavior if future editors aren't careful.  I can't think of a better architecture though.  I see why we need to batch these.

Reply:
I am sorry that I did not understand the purpose of this code, can you guys explain what were batched? thanks :)

Reply:
Because Is possible to modify this value on `widget.onChanged`?

Reply:
Yeah `onChanged` and the listeners registered to `TextEditingController` both may update the editing value. 

Should I call it `beginBatchEdit` & `endBatchEnd`, and allow nesting?